src/util/std_types.h

```diff
@@ -1671,4 +1670,71 @@ inline complex_typet &to_complex_type(typet &type)
   return static_cast<complex_typet &>(type);
 }
 
+class mathematical_function_typet:public typet
+{
+public:
+  mathematical_function_typet():typet(ID_mathematical_function)
+  {
+    subtypes().resize(2);
+  }
+
+  // the domain of the function is composed of zero, one, or
+  // many variables
+  class variablet:public irept
+  {
+  public:
+    // the identifier is optional
+    irep_idt get_identifier() const
+    {
+      return get(ID_identifier);
+    }
+
+    void set_identifier(const irep_idt &identifier)
+    {
+      return set(ID_identifier, identifier);
+    }
+
+    typet &type()
+    {
+      return (typet &)add(ID_type);
+    }
+
+    const typet &type() const
+    {
+      return (const typet &)find(ID_type);
+    }
+  };
+
+  using domaint=std::vector<variablet>;
+
+  domaint &domain()
+  {
+    return (domaint &)subtypes()[0].subtypes();
```

Comment:
`static_cast`

Reply:
That doesn't work, the types are incomparable.